File: src/main/java/com/core/base/corebase/client/dto/GoogleDto.java
package com.core.base.corebase.client.dto;

import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.format.annotation.DateTimeFormat;

import java.time.LocalDateTime;
import java.util.List;

public class GoogleDto {

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public static class GoogleTokenReq {
        private String code;
        private String clientId;
        private String clientSecret;
        private String redirectUri;
        private String grantType;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public static class GoogleTokenRefreshReq {
        private String clientId;
        private String clientSecret;
        private String refreshToken;
        private String grantType;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public static class GoogleTokenRes {
        private String accessToken;
        private int expiresIn;
        private String tokenType;
        private String scope;
        private String refreshToken;
        private String idToken;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public static class GoogleInfoRes {
        private String sub;
        private String name;
        private String givenName;
        private String familyName;
        private String picture;
        private String email;
        private String emailVerified;
        private String locale;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public static class GoogleTokenRefreshRes {
        private String accessToken;
        private String expiresIn;
        private String tokenType;
        private String scope;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class GoogleCalendarRes {
        private String kind;
        private String etag;
        private String summary;
        private String updated;
        private String timeZone;
        private String accessRole;
        private List<GoogleCalendarDefaultReminderRes> defaultReminders;
        private List<GoogleCalendarItemRes> items;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class GoogleCalendarDefaultReminderRes {
        private String method;
        private int minutes;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class GoogleCalendarItemRes {
        private String kind;
        private String etag;
        private String id;
        private String status;
        private String htmlLink;

        @DateTimeFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
        private LocalDateTime created;

        @DateTimeFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
        private LocalDateTime updated;

        private String summary;
        private String location;
        private GoogleCalendarItemPeopleRes creator;
        private GoogleCalendarItemPeopleRes organizer;
        private GoogleCalendarItemTimeRes start;
        private GoogleCalendarItemTimeRes end;
        private String recurringEventId;
        private GoogleCalendarItemTimeRes originalStartTime;
        private String iCalUID;
        private Integer sequence;
        private List<GoogleCalendarItemAttendeeRes> attendees;
        private String hangoutLink;
        private Object conferenceData;
        private GoogleCalendarItemRemindersRes reminders;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class GoogleCalendarItemPeopleRes {
        private String email;
        private String displayName;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class GoogleCalendarItemTimeRes {
        private String dateTime;
        private String timeZone;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class GoogleCalendarItemAttendeeRes {
        private String email;
        private String displayName;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class GoogleCalendarItemRemindersRes {
        private boolean useDefault;
        private List<Object> overrides;
    }
}

File: src/main/kotlin/com/core/base/corebase/client/dto/GoogleDto.kt
package com.core.base.corebase.client.dto

import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.databind.annotation.JsonNaming
import org.springframework.format.annotation.DateTimeFormat
import java.time.LocalDateTime

object GoogleDto {

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
    class GoogleTokenReq(
        val code: String,
        val clientId: String,
        val clientSecret: String,
        val redirectUri: String,
        val grantType: String
    )

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
    class GoogleTokenRefreshReq(
        val clientId: String,
        val clientSecret: String,
        val refreshToken: String,
        val grantType: String
    )

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
    class GoogleTokenRes(
        val accessToken: String,
        val expiresIn: Int,
        val tokenType: String,
        val scope: String,
        val refreshToken: String,
        val idToken: String?
    )

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
    class GoogleInfoRes(
        val sub: String,
        val name: String,
        val givenName: String,
        val familyName: String,
        val picture: String,
        val email: String,
        val emailVerified: String,
        val locale: String?
    )

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
    class GoogleTokenRefreshRes(
        val accessToken: String,
        val expiresIn: String,
        val tokenType: String,
        val scope: String
    )

    /*
    * 캘린더 DTO들에서 빠진 속성들이 많고 무조건 써야하는 필드 빼고는 대다수를 nullable하게 처리함
    * 참고: https://developers.google.com/calendar/api/v3/reference/events/list
     */
    class GoogleCalendarRes(
        val kind: String?,
        val etag: String?,
        val summary: String?,
        val updated: String?,
        val timeZone: String?,
        val accessRole: String?,
        val defaultReminders: List<GoogleCalendarDefaultReminderRes>?,
        val items: List<GoogleCalendarItemRes>
    )

    class GoogleCalendarDefaultReminderRes(
        val method: String,
        val minutes: Int
    )

    class GoogleCalendarItemRes(
        val kind: String?,
        val etag: String?,
        val id: String,
        val status: String?,
        val htmlLink: String?,
        @DateTimeFormat(pattern = "yyyy-MM-ddTHH:mm:ss.SSSZ")
        val created: LocalDateTime?,
        @DateTimeFormat(pattern = "yyyy-MM-ddTHH:mm:ss.SSSZ")
        val updated: LocalDateTime?,
        val summary: String?,
        val location: String?,
        val creator: GoogleCalendarItemPeopleRes?,
        val organizer: GoogleCalendarItemPeopleRes?,
        val start: GoogleCalendarItemTimeRes,
        val end: GoogleCalendarItemTimeRes,
        val recurringEventId: String?,
        val originalStartTime: GoogleCalendarItemTimeRes?,
        val iCalUID: String?,
        val sequence: Int?,
        val attendees: List<GoogleCalendarItemAttendeeRes>?,
        val hangoutLink: String?,
        val conferenceData: Any?,
        val reminders: GoogleCalendarItemRemindersRes?,
        val eventType: String?
    )

    class GoogleCalendarItemPeopleRes(
        val email: String,
        val displayName: String?
    )

    class GoogleCalendarItemTimeRes(
        val date: String?,
        val dateTime: String?,
        val timeZone: String?
    )

    class GoogleCalendarItemAttendeeRes(
        val email: String,
        val displayName: String?
    )

    class GoogleCalendarItemRemindersRes(
        val useDefault: Boolean
    )

}